### PR TITLE
DFM #4 Trainer, Dataset, Losses, Distance Computation, Cuda

### PR DESCRIPTION
**Pull request by gviga**
This PR is the last PR to complete the first design of the learning module of our library.
This is a massive PR that implements different modules.

DATASET: 
1) We need a way to store the entire dataset onto which to train the models. For this reason, we store the shapes in dictionaries that contain:
- mesh class
- corr (a gt correspondence for evaluation if it exists)
- distances (distance matrices) (this could also be included in the mesh class, but in this first design, they are separated)

2) Once we have a dataset with shapes, we create a Pair dataset to create a dataset of pairs of shapes.

LOSSES:
This implements a lot of losses that can be used in functional map training and evaluation. Each loss assumes to have as input specific arguments that are specified in the corpus of the class. Then we instantiate a loss manager that computes the sum of the losses onto which optimize parameters.

TRAINER:
We provide a trainer that takes as input: 
Model, PairDataset for training, PairDataset for validation, Loss Manager with specified losses, Optimizer, and Optimizer.

Notes:
1) Since we need to use CUDA, we need to add some to_device functions inside the code. This makes the overall coda a bit messy, but I don't know how to handle this better.
2) We provide an Implementation of distance computation based on Scipy that is faster than the previous one.

### CODE REVIEW

File: geomfum/learning/losses.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    required_inputs = ["fmap12", "fmap21"]
+
+    def forward(self, fmap12, fmap21):
+        """Forward pass."""
+        metric = SquaredFrobeniusLoss()
+        eye = torch.eye(fmap12.shape[0], device=fmap12.device)
+        return self.weight * (
+            metric(torch.mm(fmap12.T, fmap12), eye)
+            + metric(torch.mm(fmap21.T, fmap21), eye)
```
Comment (GiLonga) [inline geomfum/learning/losses.py:91]: Valid also for the following classes:
Does torch.mm works also with tensors and not only with 2D matrices?

Comment (GiLonga) [reply geomfum/learning/losses.py:91]: Because from the doc string I expect :
`fmap12 : torch.Tensor
        Functional map tensor of shape (batch_size, dim_out, dim_in).`

File: geomfum/learning/losses.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    ]
+
+    def _compute_geodesic_loss(self, p2p, target_dist, source_corr, target_corr):
+        """Compute the geodesic loss for batched inputs."""
+        return torch.mean(target_dist[p2p[source_corr], target_corr])
+
+    def forward(self, p2p12, dist_b, corr_a, corr_b):
+        """Forward pass."""
+        loss = self._compute_geodesic_loss(p2p12, dist_b, corr_a, corr_b)
+        return loss
```
Comment (GiLonga) [inline geomfum/learning/losses.py:210]: Fix Doc strings

File: geomfum/learning/losses.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    -------
+    torch.Tensor
+        Scalar tensor representing the weighted mean squared Frobenius norm between C^T C and the identity matrix.
+    """
+
+    def __init__(self, weight=1):
+        super().__init__()
+        self.weight = weight
+
+    required_inputs = ["fmap12", "fmap21"]
```
[the remark this answers is not in the document]
Comment (gviga) [reply geomfum/learning/losses.py:83]: The [required_inputs] variable is defined as a class attribute (not an instance attribute) because it represents a static list of input keys needed by the loss function, and this list does not change between instances. This approach is consistent with other loss classes in the codebase and allows easy access to the required inputs for each loss function. The [forward] method remains an instance method, as it operates on the instance and may use instance-specific state.

File: geomfum/learning/losses.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    Returns
+    -------
+    torch.Tensor
+        Scalar tensor representing the mean squared Frobenius norm between `a` and `b`.
+    """
+
+    def forward(self, a, b):
+        """Forward pass."""
+        return torch.mean(torch.sum(torch.abs(a - b) ** 2, dim=(-2, -1)))
```
[the remark this answers is not in the document]
Comment (gviga) [reply geomfum/learning/losses.py:61]: I think that in this way this formulation in compatible both for bartched and non batched inputs

File: geomfum/descriptor/learned.py [hunk 5]
```diff
@@ -5,10 +5,12 @@
...
-import torch
+import geomstats.backend as gs
 
 from geomfum._registry import FeatureExtractorRegistry, WhichRegistryMixins
 from geomfum.descriptor._base import Descriptor
+import torch
+import torch.nn as nn
 
 
 class BaseFeatureExtractor(abc.ABC):
```
[the remark this answers is not in the document]
Comment (gviga) [reply geomfum/descriptor/learned.py:16]: ? don't undestand

File: geomfum/dataset/torch.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            else:
+                metric = ScipyGraphShortestPathMetric(mesh)
+                geod_distance_matrix = metric.dist_matrix()
+                os.makedirs(os.path.dirname(dist_path), exist_ok=True)
+                scipy.io.savemat(
+                    dist_path,
+                    {"D": gs.to_numpy(geod_distance_matrix)},
+                )
+
+            data.update({"dist_matrix": gs.array(geod_distance_matrix)})
```
[the remark this answers is not in the document]
Comment (gviga) [reply geomfum/dataset/torch.py:131]: Changed.

File: geomfum/descriptor/pipeline.py [hunk 7]
```diff
@@ -146,8 +148,12 @@ def apply(self, shape):
         descr = None
         for step in self.steps:
             if isinstance(step, Descriptor):
-                descr = self._update_descr(descr, step(shape))
-
+                if isinstance(step, LearnedDescriptor):
+                    with torch.no_grad():
+                        new = step(shape)
+                    descr = self._update_descr(descr, new)
```
[the remark this answers is not in the document]
Comment (gviga) [reply geomfum/descriptor/pipeline.py:154]: To  discuss in the future

File: geomfum/learning/losses.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    def __init__(self, losses):
+        self.losses = losses
+
+    def compute_loss(self, outputs):
+        """Compute the total loss and a dictionary of individual losses."""
+        total_loss = 0
+        loss_dict = {}
+        for loss_fn in self.losses:
+            # Get required input keys for this loss
+            required_keys = getattr(loss_fn, "required_inputs", None)
```
[the remark this answers is not in the document]
Comment (gviga) [reply geomfum/learning/losses.py:39]: Using [getattr(loss_fn, "required_inputs", None)] is a defensive programming pattern. It allows the [LossManager] to work with any loss function, even if it doesn't define a [required_inputs] attribute. This way, you can mix custom loss modules (with or without [required_inputs].

File: geomfum/dataset/torch.py
Comment (GiLonga) [inline geomfum/dataset/torch.py:1]: We expect, almost in the learning phase, that the shapes have the same number of vertices. A check should be added to enforce this constraint, along with an error message to clearly inform the user that this requirement is mandatory.

File: geomfum/shape/mesh.py [hunk 9]
```diff
@@ -237,6 +237,41 @@ def face_areas(self):
...
+                    ).flatten()
+                )
+
+            vertex_normals = vertex_normals / (
+                gs.linalg.norm(vertex_normals, axis=1, keepdims=True) + 1e-12
+            )
+
+            self._vertex_normals = vertex_normals
+
+        return self._vertex_normals
```
Comment (GiLonga) [inline geomfum/shape/mesh.py:273]: method already defined line 188
Redefinition of unused `vertex_normals` from line 188

File: notebooks/demos/Deep-Functional-Maps-Demo.ipynb
Comment (GiLonga) [inline notebooks/demos/Deep-Functional-Maps-Demo.ipynb:1]: I also suggest to remark  in the notebook that the shapes must have the same number of vertices.

**Review by GiLonga — CHANGES_REQUESTED** [review]

File: geomfum/dataset/torch.py
Comment (gviga) [reply geomfum/dataset/torch.py:1]: This is not true

Comment (gviga) [reply geomfum/dataset/torch.py:1]: The problem arise from the geodesic distance loss, i'll fix it

File: notebooks/how_to/functional_map_forward.ipynb
Comment (GiLonga) [inline notebooks/how_to/functional_map_forward.ipynb:1]: Is necessary here to set: os.environ["GEOMSTATS_BACKEND"] = "numpy"

File: geomfum/dataset/torch.py
Comment (gviga) [reply geomfum/dataset/torch.py:1]: The problem does not arise from the loss, but from the fact that in order to use the Geodesic distance loss we need to have valuable ground truth correspondence arrays.

File: notebooks/how_to/functional_map_forward.ipynb
Comment (gviga) [reply notebooks/how_to/functional_map_forward.ipynb:1]: no, its default

**Review by GiLonga — DISMISSED** [review]

**Review by GiLonga — APPROVED** [review]
